### ISSUE TITLE
SQL query in VRT using OGRSQL dialect does not filter results correctly

### ISSUE THREAD

**Issue by arredond** [OSGeo/gdal#7087]
## Expected behavior and actual behavior.

I expect an OGR SQL query specified in a VRT using `<SrcSQL dialect="OGRSQL">` to behave in the same way as passing the same query to `ogr2ogr` using `-dialect OGRSQL -sql ...`. However the latter correctly applies a geometry type filter while the former does not.

## Steps to reproduce the problem.

Testing on a simple GPKG with four rows that contains both simple polygons (3 rows) and multi-polygons (1 row). While the GPKG spec supposedly doesn't allow this, it is actually very common. Unfortunately, Github doesn't allow me to upload a GPKG here...

```bash
ogrinfo -so -al null-island-polygons.gpkg

INFO: Open of `null-island-polygons.gpkg'
      using driver `GPKG' successful.

Layer name: null-island-polygons
Geometry: Unknown (any)
Feature Count: 4
...
```

Using `ogr2ogr`, I can apply a simple filter on the geometry type to select only the three simple polygons:

```bash
ogrinfo -so -al -dialect OGRSQL -sql "SELECT * FROM \"null-island-polygons\" WHERE OGR_GEOMETRY = 'POLYGON'" null-island-polygons.gpkg

INFO: Open of `null-island-polygons.gpkg'
      using driver `GPKG' successful.

Layer name: null-island-polygons
Geometry: Unknown (any)
Feature Count: 3
...
```

However, I wish to perform this operation using a VRT. Here's what that looks like:

```xml
<OGRVRTDataSource>
    <OGRVRTLayer name="some_layer_name">
        <SrcDataSource>null-island-polygons.gpkg</SrcDataSource>
        <SrcSQL dialect="OGRSQL">
            SELECT * FROM "null-island-polygons" WHERE OGR_GEOMETRY = 'POLYGON'
        </SrcSQL>
    </OGRVRTLayer>
</OGRVRTDataSource>
```

Let's suppose that file is called `vrt_ogrsql.xml`. Now, if I inspect it directly, the filter is not applied:

```bash
INFO: Open of `vrt_ogrsql.xml'
      using driver `OGR_VRT' successful.

Layer name: some_layer_name
Geometry: Unknown (any)
Feature Count: 4
...
```

The same operations work as expected using the SQLite dialect instead. For instance, the following command and VRT both return only 3 features:

```bash
ogrinfo -so -al -dialect SQLITE -sql "SELECT * FROM \"null-island-polygons\" WHERE ST_GeometryType(geom) = 'POLYGON'" null-island-polygons.gpkg
```

```xml
<OGRVRTDataSource>
    <OGRVRTLayer name="some_layer_name">
        <SrcDataSource>null-island-polygons.gpkg</SrcDataSource>
        <SrcSQL dialect="SQLITE">
            SELECT * FROM "null-island-polygons" WHERE ST_GeometryType(geom) = 'POLYGON'
        </SrcSQL>
    </OGRVRTLayer>
</OGRVRTDataSource>
```

## Operating system

macOS Monterey Version 12.6 with M1 chip

## GDAL version and provenance

Tested on versions 3.6.0 and 3.5.2 installed using [Homebrew](https://formulae.brew.sh/formula/gdal)

**Comment by arredond** [comment]
On a related note, it's unclear to me whether `OGRSQL` or `OGR SQL` (with a space) is the right value for the dialect. The `ogr2ogr` [docs](https://gdal.org/programs/ogr2ogr.html#cmdoption-ogr2ogr-dialect) seem to point to the former but the [VRT docs](https://gdal.org/drivers/vector/vrt.html) seem to insinuate the latter.

When filtering a GPKG, I've noticed using `OGR SQL` fails both in `ogr2ogr` and via a VRT with a `no such column: OGR_GEOMETRY` error.

However, when filtering a GeoJSON using a VRT, using `OGR SQL` works but `OGRSQL` results in no filter applied (similar to what's reported in the ticket description). I suspect `OGR SQL` (with space) is actually an invalid string and OGR is just defaulting to the `OGRSQL` dialect anyway, whereas when it's explicitly stated, the filter is not applied.

**Comment by rouault** [comment]
fix in https://github.com/OSGeo/gdal/pull/7088

and this is "OGRSQL" (also just fixed per 5358ee40c2)
